File: leetcode_python/Bit_Manipulation/construct-the-minimum-bitwise-array-ii.py
"""

3315. Construct the Minimum Bitwise Array II
Medium

You are given an array nums consisting of n prime integers.

You need to construct an array ans of length n, such that, for each index i, the bitwise OR of ans[i] and ans[i] + 1 is equal to nums[i], i.e. ans[i] OR (ans[i] + 1) == nums[i].

Additionally, you must minimize each value of ans[i] in the resulting array.

If it is not possible to find such a value for ans[i] that satisfies the condition, then set ans[i] = -1.


Example 1:

Input: nums = [2,3,5,7]
Output: [-1,1,4,3]
Explanation:
For i = 0, as there is no value for ans[0] that satisfies ans[0] OR (ans[0] + 1) = 2, so ans[0] = -1.
For i = 1, the smallest ans[1] that satisfies ans[1] OR (ans[1] + 1) = 3 is 1, because 1 OR (1 + 1) = 3.
For i = 2, the smallest ans[2] that satisfies ans[2] OR (ans[2] + 1) = 5 is 4, because 4 OR (4 + 1) = 5.
For i = 3, the smallest ans[3] that satisfies ans[3] OR (ans[3] + 1) = 7 is 3, because 3 OR (3 + 1) = 7.

Example 2:

Input: nums = [11,13,31]
Output: [9,12,15]
Explanation:
For i = 0, the smallest ans[0] that satisfies ans[0] OR (ans[0] + 1) = 11 is 9, because 9 OR (9 + 1) = 11.
For i = 1, the smallest ans[1] that satisfies ans[1] OR (ans[1] + 1) = 13 is 12, because 12 OR (12 + 1) = 13.
For i = 2, the smallest ans[2] that satisfies ans[2] OR (ans[2] + 1) = 31 is 15, because 15 OR (15 + 1) = 15.


Constraints:

1 <= nums.length <= 100
2 <= nums[i] <= 10^9
nums[i] is a prime number.

"""

# V0
# IDEA : SAME BIT RULE AS LC 3314 — THE VALUES JUST GET BIGGER
#
#   x OR (x+1) turns x's lowest zero bit on and leaves everything else alone,
#   so to hit a target n we clear the bit just below n's lowest zero :
#
#       b   = position of n's lowest 0 bit
#       ans = n - 2^(b-1)
#
#   the only unreachable target is an even one, which among primes means 2.
#
#   nums[i] now reaches 10^9, so the scan runs over ~30 bits instead of ~10;
#   nothing else changes.
#
# time = O(n log max), space = O(n)
class Solution(object):
    def minBitwiseArray(self, nums):
        res = []
        for v in nums:
            if v % 2 == 0:
                res.append(-1)
                continue
            b = 0
            while (v >> b) & 1:
                b += 1
            res.append(v - (1 << (b - 1)))
        return res

File: leetcode_java/src/main/java/LeetCodeJava/BitManipulation/ConstructTheMinimumBitwiseArrayII.java
package LeetCodeJava.BitManipulation;

// https://leetcode.com/problems/construct-the-minimum-bitwise-array-ii/

/**
 *  3315. Construct the Minimum Bitwise Array II
 *  Medium
 *
 *  You are given an array nums consisting of n prime integers.
 *
 *  You need to construct an array ans of length n, such that, for each index i,
 *  the bitwise OR of ans[i] and ans[i] + 1 is equal to nums[i], i.e.
 *  ans[i] OR (ans[i] + 1) == nums[i].
 *
 *  Additionally, you must minimize each value of ans[i] in the resulting array.
 *
 *  If it is not possible to find such a value for ans[i], set ans[i] = -1.
 *
 *  Example 1:
 *    Input: nums = [2,3,5,7]
 *    Output: [-1,1,4,3]
 *
 *  Example 2:
 *    Input: nums = [11,13,31]
 *    Output: [9,12,15]
 *
 *  Constraints:
 *    1 <= nums.length <= 100
 *    2 <= nums[i] <= 10^9
 *    nums[i] is a prime number.
 */
public class ConstructTheMinimumBitwiseArrayII {

    // V0
    // IDEA: SAME BIT RULE AS LC 3314 - ONLY THE VALUE RANGE GROWS
    //       x | (x+1) turns x's lowest zero bit on and leaves the rest alone, so
    //       to hit a target n we clear the bit just below n's lowest zero:
    //           b   = position of n's lowest 0 bit
    //           ans = n - 2^(b-1)
    //       the only unreachable target is an even one, which among primes is 2.
    //       nums[i] now reaches 10^9, so the scan covers ~30 bits (still inside
    //       an int since 10^9 < 2^30); nothing else changes.
    /**
     * time = O(N * log(max))
     * space = O(N)   // the output array
     */
    public int[] minBitwiseArray(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            int v = nums[i];
            if ((v & 1) == 0) {
                res[i] = -1;
                continue;
            }
            int b = 0;
            while (((v >> b) & 1) == 1) {
                b++;
            }
            res[i] = v - (1 << (b - 1));
        }
        return res;
    }
}
